src/mono/mono/metadata/reflection.c

```diff
@@ -1967,6 +1968,135 @@ mono_identifier_unescape_info (MonoTypeNameParse *info)
 		g_list_foreach(info->nested, &unescape_each_nested_name, NULL);
 }
 
+static void
+escape_type_name (const char *identifier, GString *str)
+{
+	char *escaped = mono_identifier_escape_type_name_chars (identifier);
+	g_string_append (str, escaped);
+	g_free (escaped);
+}
+
+static void
+escape_each_nested_name (void *data, void *user_data)
+{
+	const char* nested_name = (const char*) data;
+	GString *str = (GString*) user_data;
+	g_string_append_c (str, '+');
+	escape_type_name (nested_name, str);
+}
+
+typedef struct {
+	GString *str;
+	gboolean bounded;
+} FormatModifierState;
+
+static void
+format_each_modifier (void *data, void *user_data)
+{
+	gint modifier = GPOINTER_TO_INT (data);
+	FormatModifierState *state = user_data;
+	GString *str = state->str;
+	/* Bounded arrays are represented by a -2 and then a -1. Keep track of it,
+	and append the string when we see the next modifier. */
+	if (modifier == -2) {
+		state->bounded = TRUE;
+		return;
+	}
+	else if (modifier == -1)
+		g_string_append_c (str, '*');
+	else if (modifier == 0)
+		g_string_append_c (str, '&');
+	else {
+		g_string_append_c (str, '[');
+		if (state->bounded) {
+			g_string_append_c (str, '*');
+			state->bounded = FALSE;
+		} else {
+			for (int i = 0; i < modifier - 1; i++)
+				g_string_append_c (str, ',');
+		}
+		g_string_append_c (str, ']');
+	}
+}
+
+static void
+format_type_name (MonoTypeNameParse *info, GString *str)
```

Comment:
This function is not covered by any tests, but so was the code it replaced, which is much larger.